mlir/lib/PauliFrame/Transforms/CliffordTToPauliFramePatterns.cpp

```diff
@@ -0,0 +1,447 @@
+ *     scf.yield %out_qubits_0 : !quantum.bit
+ *   } else {
+ *     scf.yield %out_qubit : !quantum.bit
+ *   }
+ *   %2 = scf.if %z_parity -> (!quantum.bit) {
+ *     %out_qubits_0 = quantum.custom "Z"() %1 : !quantum.bit
+ *     scf.yield %out_qubits_0 : !quantum.bit
+ *   } else {
+ *     scf.yield %1 : !quantum.bit
+ *   }
+ *   %3 = quantum.custom "T"() %2 : !quantum.bit
+ *   %4 = <op_that_consumes_qubit> %3 : ...
+ */
+LogicalResult convertNonCliffordGate(CustomOp op, PatternRewriter &rewriter)
+{
+    LLVM_DEBUG(llvm::dbgs() << "Applying Pauli frame protocol to non-Clifford gate: "
+                            << op.getGateName() << "\n");
+    auto loc = op->getLoc();
+    auto outQubitTypes = op.getOutQubits().getTypes();
+
+    if (outQubitTypes.size() > 1) {
+        op->emitError() << "Only single-qubit non-Clifford gates are supported";
+        return failure();
+    }
+
+    auto outQubitType = outQubitTypes[0];
+    auto inQubits = op.getInQubits();
+
+    FlushOp flushOp = rewriter.create<FlushOp>(loc, rewriter.getI1Type(), rewriter.getI1Type(),
+                                               outQubitType, inQubits[0]);
+
+    auto pauliZOutQubit = insertPauliOpsAfterFlush(rewriter, loc, flushOp);
+
+    op->setOperands(pauliZOutQubit);
+
+    return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Clifford+T to Pauli Frame Patterns
+//===----------------------------------------------------------------------===//
+
+/**
+ * @brief Rewrite pattern for Clifford+T ops -> PauliFrame
+ */
+struct CliffordTToPauliFramePattern : public OpRewritePattern<CustomOp> {
+    using OpRewritePattern::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(CustomOp op, PatternRewriter &rewriter) const override
+    {
+        auto op_enum = hashGate(op);
+        switch (op_enum) {
+        case GateEnum::I:
+            return convertPauliGate(op, rewriter, false, false);
+        case GateEnum::X:
+            return convertPauliGate(op, rewriter, true, false);
+        case GateEnum::Y:
+            return convertPauliGate(op, rewriter, true, true);
+        case GateEnum::Z:
+            return convertPauliGate(op, rewriter, false, true);
+        case GateEnum::H:
+            return convertCliffordGate(op, rewriter, CliffordGate::Hadamard);
+        case GateEnum::S:
+            return convertCliffordGate(op, rewriter, CliffordGate::S);
+        case GateEnum::CNOT:
+            return convertCliffordGate(op, rewriter, CliffordGate::CNOT);
+        case GateEnum::T:
+            return convertNonCliffordGate(op, rewriter);
+        case GateEnum::Unknown: {
+            op->emitError() << "Unsupported gate: '" << op.getGateName()
+                            << "'. Only Clifford+T gates are supported for Pauli frame conversion: "
+                            << "I, X, Y, Z, H, S, S†, T, T†, and CNOT";
+            return failure();
+        }
+        }
+        return success();
```

Comment:
If it's covered, this line is deadcode right? Maybe we can just call the unreachable if this pattern intends to cover the gate enum?

```suggestion
        llvm_unreachable("unhandled gate");
```